### DISCUSSION TITLE
Environment variables not passing through to turbo command as described in documentation

### DISCUSSION THREAD

**Discussion by cereallarceny** [vercel/turborepo#7056] [Help]
I've been trying to get `dotenv-cli` working [as per the suggested setup on Turborepo's monorepo handbook](https://turbo.build/repo/docs/handbook/environment-variables). I've got the following `package.json`:

```json
{
  "name": "xxx-company",
  "version": "0.0.0",
  "private": true,
  "engines": {
    "node": ">=18.17",
    "pnpm": ">=8.14"
},
  "scripts": {
    "build": "dotenv -- turbo run build",
    "dev": "dotenv -- turbo run db:connect dev --parallel",
    "dev:dev": "dotenv -- turbo run db:connect:dev dev --parallel",
    "dev:prod": "dotenv -- turbo run db:connect:prod dev --parallel",
    "studio": "dotenv -- turbo run db:connect db:studio --parallel",
    "studio:dev": "dotenv -- turbo run db:connect:dev db:studio --parallel",
    "studio:prod": "dotenv -- turbo run db:connect:prod db:studio --parallel",
    "lint": "dotenv -- turbo run lint",
    "test": "dotenv -- turbo run test",
    "clean": "turbo run clean && rimraf node_modules",
    "pull-envs": "vercel env pull .env",
    "prepare": "husky install",
    "preinstall": "npx only-allow pnpm"
  },
  "lint-staged": {
    "*.{js,jsx,ts,tsx,html,md}": [
      "prettier --write --ignore-path .gitignore"
    ]
  },
  "devDependencies": {
    "@types/node": "^18.17.8",
    "dotenv-cli": "latest",
    "husky": "^8.0.3",
    "lint-staged": "^15.2.0",
    "prettier": "^3.2.4",
    "rimraf": "latest",
    "turbo": "latest",
    "vercel": "latest"
  }
}
```

I get a very odd error in my terminal when running `pnpm build`, `pnpm dev`, or any of the other commands with `dotenv --`:

```bash
 ERROR  a value is required for '--remote-only [<BOOL>]' but none was supplied
  [possible values: true, false]

For more information, try '--help'.

 ELIFECYCLE  Command failed with exit code 1.
```

Of course, I could just pass the `--remote-only` flag, but then I get the same warning about the `--summarize` flag. Without attaching every flag that Turborepo's CLI offers, how can I fix this?

**A few interesting notes/things I've already tried:**

- Using `npm` or `yarn` instead returns the same result
- Using `dotenv ...` instead of `dotenv -- ...` returns the same result
- Running `pnpm dev` and any other the other commands returns the same result as `pnpm build` - it's not isolated to the `build` script
- Downgrading `dotenv-cli` and `turbo` 1-3 minor versions each returns the same result
- Restarting my editor, typescript, or any auxillery services returns the same result
- Clearing my `node_modules` folder and running `pnpm install` again returns the same result
- Running `turbo run build` works perfectly, but then it doesn't include the environment variables in the build
- Interestingly enough, running `dotenv -- pnpx cowsay hello` works perfectly fine, so I don't think this has to do with `dotenv-cli` so much as `turbo` not using its default flags correctly when unspecified

**Comment by mehulkar** [comment]
I don't see anything obviously wrong with your setup. If you don't see the same issue with running `turbo` directly, then this has something to do with the `dotenv --` separator and how additional flags are being passed (or not passed). Please open an issue and provide a simplified reproduction repo if you want someone on the core team to look more closely and try running it.

**Comment by anthonyshew** [comment]
Closing for staleness.

**Comment by RaphaelBossek** [comment]
I had the same problem in a Docker image that I use for my Turbo monorepo. The environment variables are not passed to node when I use “turbo run dev” and start all monorepo projects in parallel. The solution was to enter the variables used for the respective script (e. g. dev) in turbo.json, as described in https://turbo.build/repo/docs/crafting-your-repository/using-environment-variables#adding-environment-variables-to-task-hashes. In this example `globalEnv` and `env`:

```
{
  "$schema": "https://turbo.build/schema.json",
  "globalDependencies": ["**/.env.*local"],
  "globalEnv": ["NODE_ENV"],
  "tasks": {
    "build": {
      "dependsOn": ["^build"],
      "outputs": [".next/**", "!.next/cache/**", "dist/**"]
    },
    "lint": {},
    "dev": {
      "cache": false,
      "persistent": true,
      "env": ["IN_CONTAINER", "PORT", "HOST"]
    }
  }
}
```

**Reply by vkouk** [reply]
Hello all,

Is there something wrong with my turbo json file here?

```
{
    "$schema": "./node_modules/turbo/schema.json",
    "globalDependencies": ["**/.env.*local", ".env"],
    "globalEnv": ["NODE_ENV"],
    "tasks": {
        "dev": {
            "env": ["NODE_ENV"],
            "inputs": ["$TURBO_DEFAULT$", ".env"],
            "cache": false,
            "persistent": true
        },
        "build": {
            "env": ["NODE_ENV"],
            "dependsOn": ["^build"],
            "inputs": ["$TURBO_DEFAULT$", ".env"],
            "outputs": ["dist/**", ".angular"]
        },
        "lint": { "dependsOn": ["^lint"] },
        "test": {
            "outputs": ["coverage/**"],
            "dependsOn": ["build"]
        },
        "test:watch": {
            "cache": false,
            "persistent": true
        },
        "test:e2e": { "dependsOn": ["build"] },
        "check-types": {}
    }
}
```

I have a root `.env` file where `NODE_ENV=local` although when doing a console.log from one of my libs such as `process.env.NODE_ENV` is undefined. I can't find any other docs here, can someone assist?

**Reply by RohitLakh** [reply]
@vkouk Did you solve it?

**Reply by vkouk** [reply]
@RohitLakh I had manually to load my `.env` because I don't think turborepo uses anyhow `dotenv` behind the scenes.

**Reply by RohitLakh** [reply]
Haha, just figured out, I had to also install `dotenv` for express backend

**Reply by vkouk** [reply]
Yes could also be that case. I am using nestjs btw, but the solution should be the same more or less

**Comment by sihamouda** [comment]
Hello all
when i set `"envMode": "loose",` in the turbo.json
then i use   `"dev": "dotenv -e ./.env turbo run dev",` in the package.json
it works

**Reply by sihamouda** [reply]
it is a workaround because i have a very long .env and i don't have time to hard code all the keys inside the turbo.json

**Comment by oknoorap** [comment]
I've found the solution:

1. Install `@dotenv-run/cli` in root, e.g `pnpm add @dotenv-run/cli -D -w`
2. Change your dev script: `"dev": "dotenv-run -r ./.env -- turbo run dev --env-mode=loose"`

**Reply by angelxmoreno** [reply]
your suggestion is a bit confusing. `@dotenv-run/cli` bypasses w/e we put in the turbo.json file. It looks more like a workaround than a fix

**Comment by KuroshHusseini** [comment]
I tried fixing it by loading the path to my root .env file and it worked. Not the best solution but its a good workaround: 
```
import path from 'path'
import app from './app'
import dotenv from 'dotenv'

dotenv.config({
  path: path.join(__dirname, '../../../.env'),
})

const PORT = process.env.PORT || 3002
console.log('🚀 ~ PORT:', PORT, process.env.PORT)

app.listen(PORT, () => {
  console.log(`SP Server is running on http://localhost:${PORT}`)
})
```

**Reply by angelxmoreno** [reply]
this looks like it would only work in dev. have you tried building and deploying an app with that config?

**Reply by KuroshHusseini** [reply]
I will keep you updated once I get there.
